Input GUI: Add tab to remap controls

makotech222 opened the pull request:
Screenshot:
![2016-07-23 13_19_38-citra _ input-gui-rebase-5232d38](https://cloud.githubusercontent.com/assets/4389156/17079514/2eab93ea-50d8-11e6-972c-8a07e9ba41cb.png)

UI based on #1119 

First pull request to Citra! 

This is a simple GUI for allowing user to assign keyboard keys to Citra buttons. It has been a long time since I've written C++ code (I do C# for a living) so please let me know if there are better ways to do anything.

Input is changed by clicking on one of the available buttons. The text of the button will change to [waiting] and will block all mouse input. When user presses a key on keyboard, it will change the text to the newly pressed key. Press 'Escape' while in input waiting mode to cancel. Changes will save when user exits the menu.

Edit: New screenshot posted, with new circle modifier button. Note that my next pull request will overhaul how input is handled. This GUI may be subject to an overhaul as well.

neobrain commented:
Hi, thanks for the contribution!

Do note however that there already have been at least 2 PRs on input, and it seems the feedback on those hasn't been considered when designing this PR. There is a reason we didn't merge the first two; ignoring that isn't going to make the third one more successful.

mailwl commented:
Look at screen, and don't like "Input" - it should be Keyboard Settings or something, to split with Gamepad, what do you think?

Subv commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+void ConfigureInput::HandleClick()
+{
+    QPushButton* sender = qobject_cast<QPushButton*>(QObject::sender());
+    sender->setText("[waiting]");
```
wrap this text with tr()

Subv commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+///Save all button configurations to settings file
+void ConfigureInput::applyConfiguration()
+{
+    for (int i = 0; i < Settings::NativeInput::NUM_INPUTS - 1; ++i) {
```
Why -1?

makotech222 replied:
no mapping yet for the circle pad modifier. I may change it when i finish up my input refactor in another pull request

Kloen commented on src/citra_qt/config.cpp:
```diff
@@ -2,15 +2,12 @@
-#include <QSettings>
-#include <QString>
-#include <QStringList>
+
```
3 white lines? just one it's fine I think.

Kloen commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+        QString keyValue = GetKeyName(defaults[i].toInt());
+        input_mapping[Settings::NativeInput::Values(i)]->setText(keyValue);
+    }
+}
```
New line at end of file.

Kloen commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+
+///Restore all buttons to their default values.
+void ConfigureInput::RestoreDefaults() {
+    for (int i = 0; i < Settings::NativeInput::NUM_INPUTS - 1; ++i)
```
Put the `{` on the same line.

Kloen commented:
Please, note that there are a newer version of the PR on wich u based the UI on: https://github.com/Kloen/citra/commits/input-config

Also, as @neobrain said, take a look on the feedback that wasn't properly implemented on those old PR.

makotech222 commented:
@mailwl Eventually this screen will support remapping for gamepads as well. A combobox will be used to select the device to map to. I think it should stay as Input tab

neobrain commented:
@makotech222 What will you do if I want to map buttons from more than a single device?

makotech222 commented:
@neobrain So like keyboard and gamepad at the same time?

neobrain commented:
For instance, yes.

makotech222 commented:
I don't know of any emulator that allows that, lol. I wasn't planning to support it. Do you think its necessary? Note, in my input refactor, keyboard input will always be available, but if the main input device is a gamepad, keyboard input will be used only for something like hotkeys or other emulator related functionality.

neobrain commented:
It would be careless to design anything around the assumption that we will never support such scenarios.

emmauss commented:
@makotech222 dolphin supports multiple device mapping,

mailwl commented:
@emmauss, don't push the boy! PR just for keyboard

mailwl commented:
i like a devil! @makotech222, can you implement gamepad support from https://github.com/citra-emu/citra/issues/1851 ? It's hard to implement, but really useful, Ahahaha. Kick them ass

mailwl commented:
sorry, too much horror video

makotech222 commented:
@emmauss Dolphin allows only one active device at a time. I just tested it...

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+// Refer to the license.txt file included.
+#include "configure_input.h"
+
+ConfigureInput::ConfigureInput(QWidget *parent) :
```
asterisk goes against the type.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+{
+    ui->setupUi(this);
+
+    //Initialize mapping of input enum to UI button.
```
Comments should have a space after the `//` (this applies to all instances)

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+        /*if (event->key() == Qt::Key_Shift)
+            return;
+
+        else if (event->key() == Qt::Key_Control)
```
`else` is not necessary with a preceding return above it.

makotech222 replied:
This is commented out and will be reworked if it is ever supported.

lioncash replied:
In that case, this should be removed and introduced whenever the PR that adds support for it is created.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+    if (key_code == Qt::Key_Shift)
+        return tr("Shift");
+
+    else if (key_code == Qt::Key_Control)
```
`else` isn't necessary here. (along with the others below)

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+}
+
+///Convert letter/name of key to its ASCII value.
+int ConfigureInput::GetKeyValue(QString text)
```
This should be a signature of 

```cpp
Qt::Key ConfigureInput::GetKeyValue(const QString& text)
```

which returns `Qt::Key_unknown` for the error case

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+{
+    if (text == "Shift")
+        return Qt::Key_Shift;
+    else if (text == "Ctrl")
```
`else` isn't necessary for conditions that have a return preceding it in the previous conditional.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+}
+
+///Check all inputs for duplicate keys. Clears out any other button with same key as new button.
+void ConfigureInput::RemoveDuplicates(QString newValue)
```
`const QString&`

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+    Q_OBJECT
+
+public:
+    explicit ConfigureInput(QWidget *parent = 0);
```
```cpp
explicit ConfigureInput(QWidget* parent = nullptr);
```

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+    QPushButton* changingButton = nullptr; /// button currently waiting for key press.
+
+    void HandleClick();
+    void setConfiguration();
```
This is inconsistently cased

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+private:
+    std::unique_ptr<Ui::ConfigureInput> ui;
+    std::map<Settings::NativeInput::Values, QPushButton*> input_mapping;
+    std::vector<int> keysPressed;
```
keys_pressed.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+    std::unique_ptr<Ui::ConfigureInput> ui;
+    std::map<Settings::NativeInput::Values, QPushButton*> input_mapping;
+    std::vector<int> keysPressed;
+    QPushButton* changingButton = nullptr; /// button currently waiting for key press.
```
`changing_button`

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include <memory>
+
+#include <QKeyEvent>
```
Many of these can be a forward declaration.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
```
Should have a newline after this line.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+    void RemoveDuplicates(QString newValue);
+    void RestoreDefaults();
+    virtual void  keyPressEvent(QKeyEvent *event);
+    QString GetKeyName(int key_code);
```
```cpp
QString GetKeyName(int key_code) const;
int GetKeyValue(const QString& text) const;
```

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+    };
+
+    //Attach handle click method to each button click.
+    for (auto &ent1 : input_mapping) {
```
`const auto& entry`

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,45 @@
+    void SetKey();
+    void RemoveDuplicates(QString newValue);
+    void RestoreDefaults();
+    virtual void  keyPressEvent(QKeyEvent *event);
```
```cpp
void keyPressEvent(QKeyEvent* event) override;
```

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+///Set button text to name of key pressed.
+void ConfigureInput::SetKey()
+{
+    QString keyValue = "";
```
`key_value`

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+}
+
+///Handle key press event for input tab when a button is 'waiting'.
+void ConfigureInput::keyPressEvent(QKeyEvent *event)
```
```cpp
QKeyEvent* event
```

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,178 @@
+///Handle key press event for input tab when a button is 'waiting'.
+void ConfigureInput::keyPressEvent(QKeyEvent *event)
+{
+    if (changingButton != nullptr && event->key() > 0)
```
`Qt::Key_unknown` has a value of `0x01ffffff`

lioncash commented on src/citra_qt/config.h:
```diff
@@ -5,6 +5,10 @@
 
 #include <string>
+#include "core/settings.h"
+#include <QSettings>
```
Why are these includes here?

lioncash commented on src/citra_qt/config.h:
```diff
@@ -21,3 +25,4 @@ class Config {
     void Reload();
     void Save();
 };
+extern const std::array<QVariant, Settings::NativeInput::NUM_INPUTS> defaults;
```
Create a function that returns a const reference to the array. Don't expose something as a global if it's not necessary.

makotech222 replied:
Sorry i'm a bit confused on this one. Should the function and array be members of the Config class? My C++ is weak :(

lioncash replied:
You create a function:

```cpp
const std::array<QVariant, Settings::NativeInput::NUM_INPUTS>& GetInputDefaults() {
    return defaults;
}
```

in the cpp file and expose that instead of a global.

makotech222 commented:
Fixed most issues except for global array. will also add Escape key as way to cancel. Will also have to look at tab key not getting caught, which i thought i fixed...

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,46 @@
+
+#pragma once
+#include <memory>
+#include <QKeyEvent>
```
This can also be a forward declaration.

makotech222 replied:
Doing so caused a compiler error.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,46 @@
+class QObject;
+class QPushButton;
+class QString;
+class QWidget;
```
QWidget needs to be an include because it's being concretely used in this file (e.g. the class inherits from it).

makotech222 replied:
Yes please. I'm really rusty on how C++ linking works. One of my least favorite things about C++

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,46 @@
+#include "core/settings.h"
+#include "ui_configure_input.h"
+
+class QObject;
```
QObject isn't used concretely or as a definition in here, so this can be removed.

lioncash commented on src/citra_qt/config.cpp:
```diff
@@ -1,221 +1,218 @@
+// Copyright 2014 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
```
Why is the whole file diffed?

makotech222 replied:
I have no idea. Ran codemaid cleanup and untabify on it.

lioncash commented on src/citra_qt/config.h:
```diff
@@ -1,23 +1,25 @@
+
+#pragma once
+
+#include "core/settings.h"
```
uh, the `<string>` include should still exist here.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,177 @@
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+#include "configure_input.h"
+
```
`<utility>` should be included here for make_pair. (standard includes come before regular project includes, as it's the general convention in other files).

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,177 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
```
Should be a newline after this line.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,177 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+#include "configure_input.h"
```
`#include "citra_qt/configure_input.h"`

lioncash commented on src/citra_qt/configure_dialog.cpp:
```diff
@@ -27,4 +27,5 @@ void ConfigureDialog::applyConfiguration() {
     ui->generalTab->applyConfiguration();
     ui->audioTab->applyConfiguration();
     ui->debugTab->applyConfiguration();
+    ui->inputTab->applyConfiguration();
```
This line should come after the general tab to match the layout.

lioncash commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,177 @@
+        return tr("Alt");
+
+    if (key_code == Qt::Key_Meta)
+        return "";
```
newline after this to be consistent.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,46 @@
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
```
Newline after the pragma, general convention, etc, etc.

emmauss commented:
@makotech222 .yes one active device. but if you middle click on a button, you can set that button to use another device. pcsx2 and ppsspp can do similar things.

makotech222 commented:
Okay more cleanup done. Thank you @lioncash for taking the time to go through my code, I appreciate it.

@emmauss Ah i see it now. On dolphin, it is right click, and you can assign from different devices. My mistake. I will take into consideration when I do input refactor.

Kloen commented:
@makotech222 Tab button will trigger default OS behavior and switch focus between elements. I think there are a workaround for that out there. I remember facing that issue when working on the old PR.

makotech222 commented:
Anyone know why I'm getting trailing white space build error in travis CI? I don't see it anywhere in my code.

Kloen commented:
@makotech222 That's because that files had a space on the end of each printed lines, you can see it with visual studio, on Edit -> Advanced -> Watch white spaces. An easy way of remove all of them is open that files with Notepad++ and go to Edit -> Clean operations -> Remove trailing witespace (Or whatever it looks on an english version of Notepad++). Another option is manually remove the latest space of each line.

makotech222 commented:
@Kloen  I do have it enabled, and I don't see whitespace in any of my files.

Kloen commented:
@makotech222 definetly there are something weird with whitespaces, just look at file diff with or without whitespaces (https://github.com/citra-emu/citra/pull/1900/files vs https://github.com/citra-emu/citra/pull/1900/files?w=1)

makotech222 commented:
I'm running Notepad++ trim trailing whitespace, and nothing is changing. There isn't any whitespace on my files >.>

makotech222 commented:
Is it possible that my line-ending setting is wrong for visual studio? Should it be LF? CL+LF? None?

makotech222 commented:
Okay I think i've fixed the issues, aside from the tab key not getting caught. I looked at Kloen's old PR and couldn't find anything that might fix it. Also searched google for an answer and found nothing. Not sure what to do on that front. Are there any other issues that anyone has with this implementation?

lioncash commented:
One thing I forgot to ask about. If someone changes their controls, but hits cancel, will their settings revert to what they initially had when opening the dialog?

lioncash commented on src/citra_qt/config.h:
```diff
@@ -20,4 +24,5 @@ class Config {
 
     void Reload();
     void Save();
+    static const std::array<QVariant, Settings::NativeInput::NUM_INPUTS>& getDefaultInput();
```
`GetDefaultInput`

makotech222 commented:
It will revert to what it was before pressing the button to change. Not necessarily what it was before opening the dialog.

wwylele commented:
Here are problems I found when trying it. Some of them is already talked on IRC. I just put it here as a reminder. The system is Ubuntu 16.04. I write these more as an end user rather than a developer.

 - Layout problem
![screenshot from 2016-06-16 08-05-31](https://cloud.githubusercontent.com/assets/4592895/16106640/f166b376-339e-11e6-97e2-ed47d7a31001.png)
You can try resizing the config window and you will see these.

 - The input configure seems allowing multi-mapping (i.e. map one keyboard key to two or more 3DS button), but it doesn't actually work. If multi-mapping cannot be implemented, then you should forbid it, or at least give a warning (such as high lighting the button that is multi-mapped)

 - When I click a button and it is waiting for my keyboard input, my mouse input is blocked - nothing can receive my mouse input, even other program can't. I also don't find an easy way to cancel key setting (if I mis-push a button) - I know I can cancel it by pressing ESC, but isn't that so hidden? What I expected is, if I press a wrong button, I can just click my mouse somewhere else to let the button lose focus and cancel it.

makotech222 commented:
@wwylele Is that with the latest changes? I redid UI using horizontal and vertical layout boxes a couple of commits ago. I don't know why its so radically different on your system. I thought QT was supposed to be consistent >.>

makotech222 commented:
Re: mouse input

Dolphin does the same thing, but will timeout after 5 seconds if no key is pressed. Would adding the timeout be okay for you?

wwylele commented:
I was and am testing the latest changes. This time I switch to Windows
![default](https://cloud.githubusercontent.com/assets/4592895/16169572/60d4e86a-353c-11e6-8a65-7dcc4279cc5d.PNG)

In Qt, you usually don't hardcode position / size / minimum size, but set size policy, wrap all widgets in many levels of layouts (not blank widgets!) and let Qt handle the position and size itself.

Your current UI is
![default](https://cloud.githubusercontent.com/assets/4592895/16169598/4d7f0a06-353d-11e6-84e6-cb16ddd04736.PNG)

and in a good practice it should be like ([.ui file here](https://gist.github.com/wwylele/460e797d7c2ee940dad8a3d995222ae6))
![default](https://cloud.githubusercontent.com/assets/4592895/16169603/69aacaf8-353d-11e6-9ae5-feea99e1e51e.PNG)
Note that this is not the best, just an example.

You can take a look how other configure tabs are designed. And always try to "play and break" your UI after you design it (resize it, change system font, change window style, etc...) and you will find problems that you are not aware when designing.

makotech222 commented:
@wwylele Okay I put your UI file in and tested it. Everything seems to work fine on my end, but, so did my old version. All I have is Win10 machine to test on, so i'm a bit limited in that respect. Thank you for the help!

wwylele commented:
@makotech222 my last test was also on a normal Windows 10. Did you try to drag the border of the configure window when running? 

Also why did you change the indention in the last commit?

makotech222 commented:
Yeah i changed window size in all directions. I ran codemaid clean up and change line endings to Unix, because of trailing white space issues. Also, did you see my comment about using a timer when editing key mapping? Thoughts?

makotech222 commented:
Okay, I've added a 5 second timeout. When you click on a button to remap, if you wait 5 seconds it will revert to previous mapping.

kreedz1 commented:
@makotech222  this is the cemu Emulator (wii U) input settings
![5f332574d4c91847ca1f3b1f2fa3c7f0](https://cloud.githubusercontent.com/assets/7191061/16251934/164f13c2-3822-11e6-9567-2b6e0e969936.png)

I personally would love a button mapping save option and a choice of device, that way you choose a controller or a mouse and you save your settings, so even after the next version of the emulator is there you don't have to remap the buttons or look for the file that has it on.

makotech222 commented:
@kreedz1 The button mapping will always save when you close the configuration dialog. No need for a save button really. Device choice will come in a later pull request when i implement SDL Controller support. 

A question for other devs: In the future, I will probably need more space on the GUI tab to include UI elements for device choice, deadzone, range and maybe some other things. Would increasing the width of the config dialog be okay in the future?

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,50 @@
+    ~ConfigureInput();
+    void applyConfiguration();
+
+    public Q_SLOTS:
```
```cpp
public slots:
```

this should also be one indent less.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,50 @@
+    void keyPressEvent(QKeyEvent* event) override;
+    QString getKeyName(int key_code) const;
+    Qt::Key getKeyValue(const QString& text) const;
+    private Q_SLOTS:
```
```cpp
private slots: 
```

this should also be one indent less

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,50 @@
+    int key_pressed;
+    QPushButton* changing_button = nullptr; /// button currently waiting for key press.
+    QString previous_mapping;
+    QTimer* timer = new QTimer();
```
the `new QTimer` should be in the constructor as `new QTimer(this);`, then you can get rid of the delete within the destructor.

makotech222 commented:
Thanks @lioncash. Changes for those are now checked in.

lioncash commented:
I had specifically said to put it in the constructor, particularly because the QTimer include can be a forward declaration in the header file that way, making the include only necessary in the .cpp file.

lioncash commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,49 @@
+    QString getKeyName(int key_code) const;
+    Qt::Key getKeyValue(const QString& text) const;
+    void setKey(); 
+private Q_SLOTS:
```
`private slots:`

John-Gee commented:
I'm only a user but I really like the new PCSX2 input config:
http://pcsx2.net/images/stories/frontend/progress_reports/q2-2016/onepad_new.png

It's so much clearer than just the names of the buttons... but it may be out of topic.

kreedz1 commented:
@makotech222 tested this 
![2740fc51d513b61b9741ed5d72157413](https://cloud.githubusercontent.com/assets/7191061/16545223/1caa81e0-411c-11e6-8aae-c4b77a58830a.png)
i thought it would have been better if the design looked like this 
![20140131202108 nintendo_3ds_button_map](https://cloud.githubusercontent.com/assets/7191061/16545233/732ff86a-411c-11e6-81e4-58ebe8fb174e.png) and you just click and bind

Even though it's effective and works well, it looked kinda plain and lacking the option to save the custom mapping made into a file and save it where ever i want.

makotech222 commented:
I agree it looks nice, but I'd like to keep the complexity to a minimum at the moment. Another pull request can redesign the UI if needed.

kreedz1 commented:
@makotech222 then why isn't this merged yet ? it works perfectly fine

Leo626 commented:
I'm not sure if this is related to Scaled Resoultion, but I noticed it looks worse with it when screens are swapped in Large Screen layout. I test this with games that primarily use the bottom screen.

**Scaled Resolution On**
![citra-qt_2016-07-03_12-38-15](https://cloud.githubusercontent.com/assets/17358613/16546537/f0b79de8-411b-11e6-8bfc-d6e37252d791.png)
**Scaled Resolution Off**
![citra-qt_2016-07-03_12-38-25](https://cloud.githubusercontent.com/assets/17358613/16546542/05504318-411c-11e6-9560-7131d705f27d.png)

wwylele commented on src/citra_qt/config.h:
```diff
@@ -3,15 +3,19 @@
 // Refer to the license.txt file included.
 
 #pragma once
-
```
I think our convention is to keep this blank line

wwylele commented on src/citra_qt/configure.ui:
```diff
@@ -10,6 +10,12 @@
+       <width>441</width>
+       <height>0</height>
+     </size>
+   </property>
```
minimumSize is not needed anymore.

wwylele commented on src/citra_qt/configure.ui:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
meaningless diff

bunnei replied:
+1

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,173 @@
+
+/// Event handler for all button released() event.
+void ConfigureInput::handleClick()
+{
```
citra's convention is no new line before brackets
`void ConfigureInput::handleClick() {`

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,173 @@
+{
+}
+
+/// Event handler for all button released() event.
```
Documents usually go in header files, I think?

makotech222 replied:
Not from what i've seen in the code base. Especially function documentation.

Jhno591 commented:
how to pull it nb thanks

archshift commented:
Before @makotech222 takes the time to rebase, does anybody else have further objections or feedback? I think we should try to get this merged.

makotech222 commented:
Okay i think this is ready. I may be doing some overhaul for the input gui in a later pull request, too, so more changes can be done there. I will be making an RFC pull request after this is accepted, for the input overhaul I did.

makotech222 commented:
Okay i rebased. Took me about 3 hours to fix git issues, and i messed up the commit name, but oh well. Everything should be good to go now.

wwylele commented on src/citra_qt/config.cpp:
```diff
@@ -52,9 +54,9 @@ void Config::ReadValues() {
-    Settings::values.bg_red   = qt_config->value("bg_red",   1.0).toFloat();
+    Settings::values.bg_red = qt_config->value("bg_red", 1.0).toFloat();
     Settings::values.bg_green = qt_config->value("bg_green", 1.0).toFloat();
-    Settings::values.bg_blue  = qt_config->value("bg_blue",  1.0).toFloat();
```
don't change these. These should go to other format clean up PR

makotech222 replied:
I personally believe we should just let codemaid handle cleanup. If we make special exceptions to align lines, then we can't use it on the rest of the file for fixing other formatting issues. Alas, i will make the change for now

wwylele commented on src/citra_qt/config.cpp:
```diff
@@ -144,9 +146,9 @@ void Config::SaveValues() {
-    qt_config->setValue("bg_red",   (double)Settings::values.bg_red);
+    qt_config->setValue("bg_red", (double)Settings::values.bg_red);
     qt_config->setValue("bg_green", (double)Settings::values.bg_green);
-    qt_config->setValue("bg_blue",  (double)Settings::values.bg_blue);
```
ditto

wwylele commented on src/citra_qt/config.h:
```diff
@@ -1,17 +1,22 @@
-// Copyright 2014 Citra Emulator Project
```
is this a line break change?

makotech222 replied:
No idea. might've messed something up during rebase?

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,163 @@
+    this->setConfiguration();
+}
+
+ConfigureInput::~ConfigureInput(){
```
space before `{`

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,163 @@
+        QString keyValue = getKeyName(Config::GetDefaultInput()[i].toInt());
+        input_mapping[Settings::NativeInput::Values(i)]->setText(keyValue);
+    }
+}
```
new line at the end of file. Also for other files

wwylele commented on src/citra_qt/config.h:
```diff
@@ -1,17 +1,22 @@
     QSettings* qt_config;
     std::string qt_config_loc;
 
+    static const std::array<QVariant, Settings::NativeInput::NUM_INPUTS> defaults;
```
actually this can be public since its a const, and we can get rid of its getter function.

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,163 @@
+/// Handle key press event for input tab when a button is 'waiting'.
+void ConfigureInput::keyPressEvent(QKeyEvent* event){
+    if (changing_button != nullptr && event->key() != Qt::Key_unknown)
+    {
```
no new line before `{`

wwylele commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,163 @@
+    if (text == "Meta")
+        return Qt::Key_unknown;
+    if (text == "")
+        return Qt::Key_unknown;
```
`getKeyName` and `getKeyValue` use different code style (one has blank lines between `if`s and one doesn't). I am not sure which is good, but at least they should be unified

wwylele commented on src/citra_qt/config.cpp:
```diff
@@ -107,9 +109,9 @@ void Config::ReadValues() {
-                                                                           qt_config->value("Context").toInt())));
+                UISettings::Shortcut(group + "/" + hotkey,
+                    UISettings::ContextualShortcut(qt_config->value("KeySeq").toString(),
+                        qt_config->value("Context").toInt())));
```
ditto

wwylele commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,49 @@
+    std::unique_ptr<Ui::ConfigureInput> ui;
+    std::map<Settings::NativeInput::Values, QPushButton*> input_mapping;
+    int key_pressed;
+    QPushButton* changing_button = nullptr; /// button currently waiting for key press.
```
If it is a doc comment I think it should be `//< `? I am not sure about this, needs others to confirm.

bunnei replied:
It should be "///<", which is for commenting things preceding the comment. 
"///" is for commenting things that follow the comment.

Jhno591 commented:
Very nice all keep up the good work :) ^

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-07-23T15:07:37Z: https://github.com/citra-emu/citra/compare/624cd9cd983e294cea77a337e8aed916f28ababe...makotech222:70800b8babb290a3d4e0552bc48d3bcdc4b16db1

makotech222 commented:
Okay fixed most of issues, except for the doc comment one.

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-07-23T18:09:17Z: https://github.com/citra-emu/citra/compare/624cd9cd983e294cea77a337e8aed916f28ababe...makotech222:5232d382ef30621661357c616467cf64798e08b6

wwylele commented:
except for the two meaningless diff above (probably not important), LGTM

bunnei commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,48 @@
+
+#pragma once
+
+#include <memory>
```
What do you need this include for? If it's something in the .cpp, then move this to there.

bunnei commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,48 @@
+    QString getKeyName(int key_code) const;
+    Qt::Key getKeyValue(const QString& text) const;
+    void setKey();
+    private slots:
```
This should be moved one indentation to the left (get rid of the spaces)

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+/// Restore all buttons to their default values.
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    return Qt::Key(QKeySequence(text)[0]);
+}
+
+/// Check all inputs for duplicate keys. Clears out any other button with same key as new button.
```
Docstrings should appear in the header file, not in the .cpp

bunnei replied:
nit: Grammar
"Clears out any other button with the same value as this button's new value."

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+void ConfigureInput::removeDuplicates(const QString& newValue) {
+    for (int i = 0; i < Settings::NativeInput::NUM_INPUTS; ++i) {
+        if (changing_button != input_mapping[Settings::NativeInput::Values(i)]) {
+            QString oldValue = input_mapping[Settings::NativeInput::Values(i)]->text();
```
This can be `const`

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+/// Restore all buttons to their default values.
+void ConfigureInput::restoreDefaults() {
+    for (int i = 0; i < Settings::NativeInput::NUM_INPUTS; ++i) {
+        QString keyValue = getKeyName(Config::defaults[i].toInt());
```
This can be `const`

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    return QKeySequence(key_code).toString();
+}
+
+/// Convert letter/name of key to its ASCII value.
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    previous_mapping = nullptr;
+}
+
+/// Convert key ASCII value to its' letter/name
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+/// Set button text to name of key pressed.
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+/// Handle key press event for input tab when a button is 'waiting'.
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    Settings::Apply();
+}
+
+/// Load configuration settings into button text
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+ConfigureInput::~ConfigureInput() {
+}
+
+/// Event handler for all button released() event.
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    this->setConfiguration();
+}
+
+ConfigureInput::~ConfigureInput() {
```
No need for this -- Default destructor

bunnei commented on src/citra_qt/configure_input.h:
```diff
@@ -0,0 +1,48 @@
+
+public:
+    explicit ConfigureInput(QWidget* parent = nullptr);
+    ~ConfigureInput();
```
No need for this -- Default destructor

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+    timer->start(5000); //Cancel after 5 seconds
+}
+
+/// Save all button configurations to settings file
```
Docstrings should appear in the header file, not in the .cpp

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+
+ConfigureInput::ConfigureInput(QWidget* parent) :
+    QWidget(parent),
+    ui(new Ui::ConfigureInput) {
```
nit: Put all of these on one line, it looks silly this way

lioncash replied:
This particular line can also use `std::make_unique` instead of `new`

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+        { std::make_pair(Settings::NativeInput::Values::DDOWN, ui->btnDirDown) },
+        { std::make_pair(Settings::NativeInput::Values::DLEFT, ui->btnDirLeft) },
+        { std::make_pair(Settings::NativeInput::Values::DRIGHT, ui->btnDirRight) },
+        { std::make_pair(Settings::NativeInput::Values::CUP, ui->btnStickUp) },
```
The choice of name here is a little ambiguous IMO... should these be "buttonCStickUp"?

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    // Initialize mapping of input enum to UI button.
+    input_mapping = {
+        { std::make_pair(Settings::NativeInput::Values::A, ui->btnFaceA) },
```
nit: avoid pointlessly dropping characters... "button" is more readable then "btn"

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+        { std::make_pair(Settings::NativeInput::Values::B, ui->btnFaceB) },
+        { std::make_pair(Settings::NativeInput::Values::X, ui->btnFaceX) },
+        { std::make_pair(Settings::NativeInput::Values::Y, ui->btnFaceY) },
+        { std::make_pair(Settings::NativeInput::Values::L, ui->btnShdrL) },
```
nit: avoid pointlessly dropping characters... "Shoulder" is more readable then "Shdr"
... FWIW, I think you can drop "Face" and "Shdr" altogether, it doesn't really add any value since the letters are unique

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+        { std::make_pair(Settings::NativeInput::Values::START, ui->btnStart) },
+        { std::make_pair(Settings::NativeInput::Values::SELECT, ui->btnSelect) },
+        { std::make_pair(Settings::NativeInput::Values::HOME, ui->btnHome) },
+        { std::make_pair(Settings::NativeInput::Values::DUP, ui->btnDirUp) },
```
The choice of name here is a little ambiguous IMO... should these be "buttonDPadUp"?

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+}
+
+/// Save all button configurations to settings file
+void ConfigureInput::applyConfiguration() {
```
saveConfiguration?

makotech222 replied:
All other tabs use setConfiguration and applyConfiguration. Just following convention

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+}
+
+/// Load configuration settings into button text
+void ConfigureInput::setConfiguration() {
```
loadConfiguration?

makotech222 replied:
same as above. following convention

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+
+/// Handle key press event for input tab when a button is 'waiting'.
+void ConfigureInput::keyPressEvent(QKeyEvent* event) {
+    if (changing_button != nullptr && event->key() != Qt::Key_unknown) {
```
* Don't need to explicitly test against `nullptr`, the boolean test is more readable, less prone to typos
* Should perform the same test on `event` before dereferencing it
* Prefer return-early code.
E.g.:
<pre>
    /// Handle key press event for input tab when a button is 'waiting'.
    void ConfigureInput::keyPressEvent(QKeyEvent* event) {
        if (!changing_button)
            return;

        if (!event || event->key() == Qt::Key_unknown)
            return;

        key_pressed = event->key();
        timer->stop();
        setKey();
    }
</pre>

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,160 @@
+
+/// Set button text to name of key pressed.
+void ConfigureInput::setKey() {
+    QString key_value = getKeyName(key_pressed);
```
Can be `const`

bunnei commented:
I added a lot of comments, but they are all very minimal things that are easy to fix. I should also get time to test this tonight (hopefully). If that goes fine and you address my comments, this should be pretty much good to go IMO. Thanks!

bunnei commented:
Tested - everything works nicely!

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-07-28T23:51:47Z: https://github.com/citra-emu/citra/compare/624cd9cd983e294cea77a337e8aed916f28ababe...makotech222:4ceb57ca2ed80ea4aaa5e653252b0bc25f37e73e

makotech222 commented:
okie dokie. that should do it for bunnei's comments.

bunnei commented on src/citra_qt/configure_input.cpp:
```diff
@@ -0,0 +1,149 @@
+        const QString keyValue = getKeyName(Config::defaults[i].toInt());
+        input_mapping[Settings::NativeInput::Values(i)]->setText(keyValue);
+    }
+}
```
Missing new line

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-07-29T00:06:43Z: https://github.com/citra-emu/citra/compare/624cd9cd983e294cea77a337e8aed916f28ababe...makotech222:7f1b9bd96d07542203fe520188c9c96f797f7353

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-07-29T01:37:22Z: https://github.com/citra-emu/citra/compare/624cd9cd983e294cea77a337e8aed916f28ababe...makotech222:964689db180c17d38f0e6834789577ea30ea1be6

bunnei commented:
LGTM
